File: src/behavioral/interpreter/Interpreter.cs
namespace DesignPatterns.Behavioral.Interpreter;

// Context class
public class Context
{
    public string Data { get; set; }
}

// Abstract Expression class
public abstract class Expression
{
    public abstract void Interpret(Context context);
}

// Terminal Expression class
public class TerminalExpression : Expression
{
    private string _data;

    public TerminalExpression(string data)
    {
        _data = data;
    }

    public override void Interpret(Context context)
    {
        if (context.Data.Contains(_data))
        {
            Console.WriteLine("TerminalExpression: Interpreting '" + _data + "'");
        }
    }
}

// Non-Terminal Expression class
public class NonTerminalExpression : Expression
{
    private List<Expression> _expressions = new List<Expression>();

    public void AddExpression(Expression expression)
    {
        _expressions.Add(expression);
    }

    public override void Interpret(Context context)
    {
        foreach (var expression in _expressions)
        {
            expression.Interpret(context);
        }
    }
}

class Program
{
    static void Main()
    {
        Context context = new Context { Data = "Hello world" };

        TerminalExpression terminalExpression = new TerminalExpression("Hello");
        NonTerminalExpression nonTerminalExpression = new NonTerminalExpression();
        nonTerminalExpression.AddExpression(terminalExpression);

        nonTerminalExpression.Interpret(context);
    }
}

File: behavioral/interpreter/interpreter.py

# Context class
class Context:
    def __init__(self, data):
        self.data = data

# Abstract Expression class
class Expression:
    def interpret(self, context):
        pass

# Terminal Expression class
class TerminalExpression(Expression):
    def __init__(self, data):
        self._data = data

    def interpret(self, context):
        if self._data in context.data:
            print(f"TerminalExpression: Interpreting '{self._data}'")

# Non-Terminal Expression class
class NonTerminalExpression(Expression):
    def __init__(self):
        self._expressions = []

    def add_expression(self, expression):
        self._expressions.append(expression)

    def interpret(self, context):
        for expression in self._expressions:
            expression.interpret(context)

if __name__ == "__main__":
    context = Context("Hello world")

    terminal_expression = TerminalExpression("Hello")
    non_terminal_expression = NonTerminalExpression()
    non_terminal_expression.add_expression(terminal_expression)

    non_terminal_expression.interpret(context)
